server/src/main/java/org/opensearch/transport/InboundPipeline.java

```diff
@@ -120,41 +121,48 @@ public void handleBytes(TcpChannel channel, ReleasableBytesReference reference)
     }
 
     public void doHandleBytes(TcpChannel channel, ReleasableBytesReference reference) throws IOException {
-        channel.getChannelStats().markAccessed(relativeTimeInMillis.getAsLong());
-        statsTracker.markBytesRead(reference.length());
-        pending.add(reference.retain());
-
-        final ArrayList<Object> fragments = fragmentList.get();
-        boolean continueHandling = true;
-
-        while (continueHandling && isClosed == false) {
-            boolean continueDecoding = true;
-            while (continueDecoding && pending.isEmpty() == false) {
-                try (ReleasableBytesReference toDecode = getPendingBytes()) {
-                    final int bytesDecoded = decoder.decode(toDecode, fragments::add);
-                    if (bytesDecoded != 0) {
-                        releasePendingBytes(bytesDecoded);
-                        if (fragments.isEmpty() == false && endOfMessage(fragments.get(fragments.size() - 1))) {
+        try {
+            byte[] incomingBytes = BytesReference.toBytes(reference);
```

Comment:
I don't think we need to deal with protobuf at here: the method extract raw bytes from the stream, one the bytes are extracted, we could use header to find out the protocol (and handler) to use.

Reply:
I think this might answer this and the previous comment:
The objects for the original message and protobuf message are different. Original message is an OutboundMessage serialized to send over the wire and then deserialized as InboundMessage. For node-to-node communication with protobuf, the object is now different with NodeToNodeMessage. I am wondering if we can extract this logic out and make it more common in the later PRs since it might alter how the current decoding over the wire logic works. WDYT @reta?

Reply:
> For node-to-node communication with protobuf, the object is now different with NodeToNodeMessage. 

I though we only serializing the payload as a protobuf? In any case, I think we should not keep this logic exception driven (try protobuf, if not - try other), we should have a very straightforward way to identify protocol before.

Reply:
For [NodeToNodeMessage](https://github.com/opensearch-project/OpenSearch/pull/11910/files#diff-3b6cef54e769860b6ff96d7052208f6fe2ef8a86ea04aec232044984ba52e5dbR19) proto, the headers are now part of that proto message so its not a different object.

Reply:
> For [NodeToNodeMessage](https://github.com/opensearch-project/OpenSearch/pull/11910/files#diff-3b6cef54e769860b6ff96d7052208f6fe2ef8a86ea04aec232044984ba52e5dbR19) proto, the headers are now part of that proto message so its not a different object. For the protobuf implementation in this change, it writes the bytes according to proto for the node-to-node-message while the original implementation writes the bytes differently.

Oh I see, do you think we  could keep headers and payload separated? The benefits of that approach would be:
 a) easy protocol detection
 b) no changes for existing protocol whatsoever

One option I've in mind is to use magic bytes in the message  (fe first 4 bytes) but it would require the modification of the existing protocol (and also the small but teal chance the the magic bytes would appear in the mixed clusters)

Reply:
I definitely agree that keeping headers and payload separate gives a lot of benefits but I think the separate proto message for node-to-node communication is a big part of the benefits that we saw in terms of benchmark numbers to improve latency. There is [custom serialization](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/OutboundMessage.java#L60) happening currently to write the header and payload separately on the stream which is avoided with protobuf since the message can be converted to bytes directly and can be sent over the wire.

For the magic bytes option, I thought we could do this in upcoming changes as it requires modification to the existing protocol. But it also allows us to reap the benefits of proto bytes over the wire and once completely implemented offers a way to detect any kind of protocol.

Reply:
@saratvemulapalli @reta What do you guys think?

Reply:
@VachaShah personally, I am not be comfortable with how we are detecting the message protocol, to me this is probably the most important thing we should get right before going further. I am sadly very short on time to experiment but I think we could look if:
 - `profobuf` encoding includes prefix (magic bytes) 
 - if not, since `profobuf` message starts with a headers, have a fixed header in every message (as the magic bytes replacement)